third_party/amd/lib/TritonAMDGPUTransforms/ConvertToBufferOps.cpp

```diff
@@ -38,10 +38,259 @@ namespace ttg = mlir::triton::gpu;
 namespace tt = mlir::triton;
 
 namespace {
-template <typename F>
-bool verifyNonSmallerByAssumption(Value expr,
-                                  const DenseSet<Value> &assumptions,
-                                  F matchesOther) {
+
+constexpr int64_t kDefaultMaxTripCount = 1024;
+const std::string kConvertBufferOpsPrefix = "__amdgpuconvertbufferops.";
+const std::string kOutputRange = kConvertBufferOpsPrefix + "output_range";
+
+std::optional<int64_t> maybeGetTripCount(LoopLikeOpInterface loop) {
+  std::optional<OpFoldResult> lowerBound = loop.getSingleLowerBound();
+  std::optional<OpFoldResult> upperBound = loop.getSingleUpperBound();
+  std::optional<OpFoldResult> step = loop.getSingleStep();
+  assert(lowerBound && upperBound && step);
+  return constantTripCount(*lowerBound, *upperBound, *step);
+}
+
+void getEnclosingLoops(Operation &op, SmallVector<LoopLikeOpInterface> &ops) {
+  Operation *currOp = op.getParentOp();
+  while (currOp) {
+    if (isa<LoopLikeOpInterface>(currOp))
+      ops.push_back(llvm::cast<LoopLikeOpInterface>(currOp));
+    currOp = currOp->getParentOp();
+  }
+}
+
+void inferResultRanges(GetProgramIdOp *op, SetIntRangeFn setResultRange) {
+  constexpr int64_t kDefaultMaxProgramID = 2048;
+  setResultRange(
+      op->getResult(),
+      ConstantIntRanges::range(
+          /*min*/ {/*numBits*/ 32, /*val*/ 0, /*isSigned*/ true},
+          /*max*/
+          {/*numBits*/ 32, /*val*/ kDefaultMaxProgramID, /*isSigned*/ true},
+          /*isSigned*/ true));
+}
+
+void inferResultRanges(MakeRangeOp *op, SetIntRangeFn setResultRange) {
+  setResultRange(
+      op->getResult(),
+      ConstantIntRanges::range(
+          /*min*/ {/*numBits*/ 32, /*val*/ op->getStart(), /*isSigned*/ true},
+          /*max*/ {/*numBits*/ 32, /*val*/ op->getEnd(), /*isSigned*/ true},
+          /*isSigned*/ true));
+}
+
+void inferResultRanges(SplatOp *op, ArrayRef<ConstantIntRanges> argRanges,
+                       SetIntRangeFn setResultRange) {
+  setResultRange(op->getResult(), argRanges[0]);
+}
+
+void inferResultRanges(ExpandDimsOp *op, ArrayRef<ConstantIntRanges> argRanges,
+                       SetIntRangeFn setResultRange) {
+  setResultRange(op->getResult(), argRanges[0]);
+}
+
+struct TritonIntegerRangeAnalysis : dataflow::IntegerRangeAnalysis {
+  using dataflow::IntegerRangeAnalysis::IntegerRangeAnalysis;
+
+  llvm::SmallDenseMap<LoopLikeOpInterface, int64_t> loopTripCounts;
+  llvm::SmallDenseMap<
+      std::pair<LoopLikeOpInterface, dataflow::IntegerValueRangeLattice *>,
+      int64_t>
+      loopVisits;
+
+  LogicalResult visitOperation(
+      Operation *op,
+      ArrayRef<const dataflow::IntegerValueRangeLattice *> operands,
+      ArrayRef<dataflow::IntegerValueRangeLattice *> results) override {
+    LDBG("  Inferring ranges for " << *op << "\n");
+    auto joinCallback = [&op, &results, this](Value v,
```

Comment:
Can you add doc? Also nit: it's better to have self-explanatory variable names. `attrs` is not a good name there; something like `newRange` is clearer.